Issue2180 boiler plant main controller oct 2021

karthikeyad-pnnl opened the pull request:
@JayHuLBL I am submitting this pull request to address the latest comments on PR #2700. Please review it when you get a chance.

Thanks,
Karthik

JayHuLBL commented:
@karthikeyad-pnnl I see following errors when run the `validateHTML.py` as below:
```
jianjunhu@ubuntu:~/GitFolder/karthikeyad-pnnl/modelica-buildings/Buildings$ ../bin/validateHTML.py 
The following malformed html syntax has been found:
[-- ./Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Pumps/PrimaryPumps/Controller.mo ]
line 53 column 1 - Info: missing optional end tag </li>

[-- ./Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Pumps/SecondaryPumps/Controller.mo ]
line 33 column 1 - Info: missing optional end tag </li>

[-- ./Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Generic/ZeroIndexCorrection.mo ]
line 65 column 7 - Warning: inserting implicit <p>
line 65 column 7 - Warning: trimming empty <p>

[-- ./Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Generic/Validation/ZeroIndexCorrection.mo ]
line 47 column 5 - Warning: inserting implicit <p>
line 47 column 5 - Warning: trimming empty <p>
```

karthikeyad-pnnl commented:
> @karthikeyad-pnnl I see following errors when run the `validateHTML.py` as below:
> 
> ```
> jianjunhu@ubuntu:~/GitFolder/karthikeyad-pnnl/modelica-buildings/Buildings$ ../bin/validateHTML.py 
> The following malformed html syntax has been found:
> [-- ./Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Pumps/PrimaryPumps/Controller.mo ]
> line 53 column 1 - Info: missing optional end tag </li>
> 
> [-- ./Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Pumps/SecondaryPumps/Controller.mo ]
> line 33 column 1 - Info: missing optional end tag </li>
> 
> [-- ./Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Generic/ZeroIndexCorrection.mo ]
> line 65 column 7 - Warning: inserting implicit <p>
> line 65 column 7 - Warning: trimming empty <p>
> 
> [-- ./Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Generic/Validation/ZeroIndexCorrection.mo ]
> line 47 column 5 - Warning: inserting implicit <p>
> line 47 column 5 - Warning: trimming empty <p>
> ```
> 
> Also when run unit test, I see some results difference as below. My Dymola version is 2023x, same as the one in CI test.
> 
> ```
> jianjunhu@ubuntu:~/GitFolder/karthikeyad-pnnl/modelica-buildings/Buildings$ ../bin/runUnitTests.py -s Buildings.Controls.OBC.ASHRAE.PrimarySystem.BoilerPlant
> Regression tests are only run for the following package:
>   Buildings.Controls.OBC.ASHRAE.PrimarySystem.BoilerPlant
> Number of models   : 2331
>           blocks   : 669
>           functions: 209
> Using 4 of 4 processors to run unit tests for dymola.
> Generated 50 regression tests.
> 
> *** Error: Buildings_Controls_OBC_ASHRAE_PrimarySystem_BoilerPlant_Staging_Processes_Subsequences_Validation_DisableBoiler.txt: Errors during result verification.
>            Absolute error = 9.990e-01 at t = 1080.0 for truFalHol.y
>              For Buildings_Controls_OBC_ASHRAE_PrimarySystem_BoilerPlant_Staging_Processes_Subsequences_Validation_DisableBoiler.txt,
>              update reference files with new trajectories?
>              Enter: y(yes), n(no), Y(yes for all), N(no for all): N
> *** Error: Buildings_Controls_OBC_ASHRAE_PrimarySystem_BoilerPlant_Staging_Processes_Validation_Up.txt: Errors during result verification.
>            Absolute error = 9.990e-01 at t = 0.0 for upProCon.uBoi[1]
>            Absolute error = 9.990e-01 at t = 108.0 for upProCon.uBoi[2]
>            Absolute error = 9.990e-01 at t = 657.0 for upProCon.uBoiSet[1]
>            Absolute error = 9.970e-01 at t = 657.0 for upProCon.uStaSet
>            Absolute error = 9.990e-01 at t = 0.0 for upProCon.uHotWatIsoVal[1]
>            Absolute error = 9.990e-01 at t = 72.0 for upProCon.uHotWatIsoVal[2]
>            Absolute error = 9.990e-01 at t = 0.0 for upProCon.yHotWatIsoVal[1]
>            Absolute error = 9.990e-01 at t = 72.0 for upProCon.yHotWatIsoVal[2]
>            Absolute error = 9.989e-01 at t = 9.0 for yPumChaPro.y
>            Absolute error = 9.989e-01 at t = 639.0 for yStaChaPro.y
>            Absolute error = 9.990e-01 at t = 0.0 for upProCon1.uBoi[1]
>            Absolute error = 9.990e-01 at t = 108.0 for upProCon1.uBoi[2]
>            Absolute error = 9.990e-01 at t = 639.0 for upProCon1.uHotWatIsoVal[1]
>            Absolute error = 9.990e-01 at t = 72.0 for upProCon1.uHotWatIsoVal[2]
>            Absolute error = 9.990e-01 at t = 639.0 for upProCon1.yHotWatIsoVal[1]
>            Absolute error = 9.990e-01 at t = 72.0 for upProCon1.yHotWatIsoVal[2]
>            Absolute error = 9.990e-01 at t = 0.0 for upProCon2.uBoi[1]
>            Absolute error = 9.990e-01 at t = 54.0 for upProCon2.uBoi[2]
>            Absolute error = 9.990e-01 at t = 558.0 for upProCon2.uBoiSet[1]
>            Absolute error = 9.970e-01 at t = 558.0 for upProCon2.uStaSet
>            Absolute error = 9.989e-01 at t = 9.0 for yPumChaPro2.y
>            Absolute error = 9.989e-01 at t = 540.0 for yStaChaPro2.y
>            Absolute error = 9.990e-01 at t = 0.0 for upProCon3.uBoi[1]
>            Absolute error = 9.990e-01 at t = 54.0 for upProCon3.uBoi[2]
>            Absolute error = 9.990e-01 at t = 0.0 for upProCon4.uBoi[1]
>            Absolute error = 9.990e-01 at t = 108.0 for upProCon4.uBoi[2]
>            Absolute error = 9.990e-01 at t = 648.0 for upProCon4.uHotWatIsoVal[1]
>            Absolute error = 9.990e-01 at t = 72.0 for upProCon4.uHotWatIsoVal[2]
>            Absolute error = 9.990e-01 at t = 648.0 for upProCon4.yHotWatIsoVal[1]
>            Absolute error = 9.990e-01 at t = 72.0 for upProCon4.yHotWatIsoVal[2]
> *** Error: Buildings_Controls_OBC_ASHRAE_PrimarySystem_BoilerPlant_Staging_SetPoints_Subsequences_Validation_Change.txt: Errors during result verification.
>            Absolute error = 9.988e+00 at t = 9240.0 for cha.uAvaUp
>            Absolute error = 8.990e+00 at t = 9240.0 for cha.uAvaDow
>            Absolute error = 9.989e+00 at t = 9240.0 for cha.ySta
>            Absolute error = 2.995e+00 at t = 8680.0 for cha2.uAvaUp
>            Absolute error = 1.997e+00 at t = 8680.0 for cha2.uAvaDow
>            Absolute error = 2.996e+00 at t = 8680.0 for cha2.ySta
> *** Error: Buildings_Controls_OBC_ASHRAE_PrimarySystem_BoilerPlant_Pumps_SecondaryPumps_Subsequences_Validation_EnableLag_pumpSpeed.txt: Errors during result verification.
>            Absolute error = 9.990e-01 at t = 3456.0 for yUp.y
> *** Error: Buildings_Controls_OBC_ASHRAE_PrimarySystem_BoilerPlant_SetPoints_Validation_MinimumFlowSetPoint.txt: Errors during result verification.
>            Absolute error = 4.590e-01 at t = 4.8 for minBoiFloSet1.VHotWatMinSet_flow
> Comparison files output by funnel are stored in the directory 'funnel_comp' of size 5.8 MB.
> Run 'python -c "import buildingspy.development.regressiontest as t; t.Tester(tool=\"dymola\").report()"'
> to access a summary of the comparison results.
> 
> Script that runs unit tests had 0 warnings and 5 errors.
> 
> See 'simulator-dymola.log' for details.
> Execution time = 294.958 s
> Server will run for 600 (s) or until KeyboardInterrupt.
> ^Z
> [9]+  Stopped                 ../bin/runUnitTests.py -s Buildings.Controls.OBC.ASHRAE.PrimarySystem.BoilerPlant
> ```

@JayHuLBL I did see the last two error instances (line 65 and line 47) when I ran `runUnitTest.py --validate-html-only`. However, I wasn't sure what I needed to do to solve it since there are no `<p>` tags that are not closed with a `</p>` tag. I did not see the first two instances. Will fix it now.

karthikeyad-pnnl commented:
@JayHuLBL Will check the unit test results. Do you have any kind of image available for the CI testing environment? I think it may be beneficial for us to run it on a container with the same image on our end. Regarding the use of Dymola 2023, did you have to update the license server when you switched from previous versions to 2023?

JayHuLBL commented:
@karthikeyad-pnnl I don't have the container for the environment. However, for the license server, I think it should be in the package along with the Dymola installation. Note that it is 2023x, not 2023.

JayHuLBL commented:
@karthikeyad-pnnl 
- [ ] In `BoilerPlant.Pumps.PrimaryPumps.Controller,missing`, it is missing`</li>` after [line 1249](https://github.com/karthikeyad-pnnl/modelica-buildings/blob/52f2f672755799c80e95999451f3791537025b3b/Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Pumps/PrimaryPumps/Controller.mo#L1249) to close the `<li>` in line 1235.
- [ ] In `BoilerPlant.Pumps.SecondaryPumps.Controller`, it is missing `</li>` after [line 866](https://github.com/karthikeyad-pnnl/modelica-buildings/blob/52f2f672755799c80e95999451f3791537025b3b/Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Pumps/SecondaryPumps/Controller.mo#L866) to close `<li>` in line 850.
- [ ] In `BoilerPlant.Generic.ZeroIndexCorrection`, move `</p>` in [line 120](https://github.com/karthikeyad-pnnl/modelica-buildings/blob/52f2f672755799c80e95999451f3791537025b3b/Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Generic/ZeroIndexCorrection.mo#L120) to be after line 109.
- [ ] In `BoilerPlant.Generic.Validation.ZeroIndexCorrection`, add `</p>` after [line 51](https://github.com/karthikeyad-pnnl/modelica-buildings/blob/52f2f672755799c80e95999451f3791537025b3b/Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Generic/Validation/ZeroIndexCorrection.mo#L51) and add `<p>` after [line 59](https://github.com/karthikeyad-pnnl/modelica-buildings/blob/52f2f672755799c80e95999451f3791537025b3b/Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Generic/Validation/ZeroIndexCorrection.mo#L59) .

karthikeyad-pnnl commented:
@AntoineGautier : I have uploaded changes to this PR to address your comments [here](https://github.com/lbl-srg/modelica-buildings/pull/2700#pullrequestreview-1384877568). Some general comments:

- I agree that the enable schedule should be a software input so that the user can configure it as per their requirements. I have changed it from a parameter to an input.
- I have changed the names of the Boolean plant configuration parameters that were ambiguous with their attached comments. I have in fact changed the top-level controller to accept parameter inputs for hardware available (sensor in primary loop, sensor in decoupler, etc.) and then use those to calculate the parameters for enabling/disabling instances.
- Regarding the ability of the controller to accommodate hybrid boiler plants: We initially developed and tested the modules with a focus on primary-only, condensing boiler plants. To make it compatible with hybrid systems, I am currently vectorizing the primary and secondary pump controllers so that they can operate each primary and secondary loop. The staging calculations will remain the same since all the boilers in the plant will still be staged by a single module (we could just use multiple controllers if the primary loops were all operating independently). Also, the staging processes will be vectorized for each loop so that the process is as per the actual sequence requirement.
These two steps should help make the controller usable for hybrid plants with multiple primary and secondary loops. (Note: The changes are still in-process. I hope to be done with them before 06/22/23).

I would like your thoughts on the changes, esp on the changes being made to accommodate hybrid plants. Thanks!

AntoineGautier commented:
@karthikeyad-pnnl @JayHuLBL I'm confused with this hanging PR.

I've recently merged https://github.com/lbl-srg/modelica-buildings/pull/4176 which merges 
- [karthikeyad-pnnl:issue2180_BoilerPlant_Nov23](https://github.com/karthikeyad-pnnl/modelica-buildings/tree/issue2180_BoilerPlant_Nov23) into 
- [lbl-srg:issue2180_BoilerPlant_MainController_oct_2021](https://github.com/lbl-srg/modelica-buildings/tree/issue2180_BoilerPlant_MainController_oct_2021)

But now I see that the recent commits from 

- [karthikeyad-pnnl:issue2180_BoilerPlant_MainController_oct_2021](https://github.com/karthikeyad-pnnl/modelica-buildings/tree/issue2180_BoilerPlant_MainController_oct_2021) 

which appear in this PR (such as [1f0b26b](https://github.com/lbl-srg/modelica-buildings/pull/3325/commits/1f0b26b29dcf03dda513976918f1e7e1a0c44700)) are not in the tree of 
- [lbl-srg:issue2180_BoilerPlant_MainController_oct_2021](https://github.com/lbl-srg/modelica-buildings/tree/issue2180_BoilerPlant_MainController_oct_2021)

Is this because some rebasing was performed?
@karthikeyad-pnnl Can you confirm that the changes from this PR are included in the new branch you created: 
- [karthikeyad-pnnl:issue2180_BoilerPlant_Nov23](https://github.com/karthikeyad-pnnl/modelica-buildings/tree/issue2180_BoilerPlant_Nov23)

which was merged via https://github.com/lbl-srg/modelica-buildings/pull/4176?

In which case, shall we close this PR and centralize all further action on this feature development in https://github.com/lbl-srg/modelica-buildings/pull/4176? 
(Before merging https://github.com/lbl-srg/modelica-buildings/pull/4176 there were 3 open PR related to the same development.)

AntoineGautier commented:
@karthikeyad-pnnl Can you confirm that this can be closed ***without*** merging?

karthikeyad-pnnl commented:
@AntoineGautier This can be closed without merging. It has some incomplete changes that are no longer relevant. The relevant changes are contained in #4195 and other previously-merged PRs
